diff --git a/src/qtils/hex.hpp b/src/qtils/hex.hpp
--- a/src/qtils/hex.hpp	(revision 38edff02e94e1a47b31896607bd4f49b454ee1da)
+++ b/src/qtils/hex.hpp	(revision 4749d1f8332e848dbce2ac74fc1a0c1992e9cbdb)
@@ -12,13 +12,16 @@
 #include <type_traits>

 namespace qtils {
+  using BytesIn = std::span<const uint8_t>;

   /**
    * @struct Hex
    * @brief Wrapper for hex encoding of byte sequences.
    */
-  struct Hex : std::span<const uint8_t> {
-    using std::span<const uint8_t>::span;
+  struct Hex : BytesIn {
+    using BytesIn::span;
+
+    Hex(BytesIn span) : BytesIn{span} {}
   };

 }  // namespace qtils
diff --git a/src/qtils/enum_error_code.hpp b/src/qtils/enum_error_code.hpp
--- a/src/qtils/enum_error_code.hpp	(revision 4749d1f8332e848dbce2ac74fc1a0c1992e9cbdb)
+++ b/src/qtils/enum_error_code.hpp	(revision 4eb3f8024817d66932cec0c52e74e127c137a78a)
@@ -8,6 +8,7 @@

 #include <system_error>
 #include <type_traits>
+#include <typeinfo>

 namespace qtils {
   template <typename E>
